pkg/cachemanager/cachemanager_test.go

```diff
@@ -0,0 +1,373 @@
+	require.NoError(t, err)
+
+	// when gvk is watched, we expect Add, Remove to work
+	cm.watchedSet.Add(pod.GroupVersionKind())
+
+	require.NoError(t, cm.AddObject(ctx, &unstructured.Unstructured{Object: unstructuredPod}))
+	require.True(t, cfClient.HasGVK(pod.GroupVersionKind()))
+
+	// now remove the object and verify it's removed
+	require.NoError(t, cm.RemoveObject(ctx, &unstructured.Unstructured{Object: unstructuredPod}))
+	require.False(t, cfClient.HasGVK(pod.GroupVersionKind()))
+
+	cm.watchedSet.Remove(pod.GroupVersionKind())
+	require.NoError(t, cm.AddObject(ctx, &unstructured.Unstructured{Object: unstructuredPod}))
+	require.False(t, cfClient.HasGVK(pod.GroupVersionKind())) // we drop calls for gvks that are not watched
+}
+
+// TestCacheManager_AddObject_processExclusion makes sure that we don't add objects that are process excluded.
+func TestCacheManager_AddObject_processExclusion(t *testing.T) {
+	cm, ctx := unitCacheManagerForTest(t)
+	processExcluder := process.Get()
+	processExcluder.Add([]configv1alpha1.MatchEntry{
+		{
+			ExcludedNamespaces: []wildcard.Wildcard{"test-ns-excluded"},
+			Processes:          []string{"sync"},
+		},
+	})
+	cm.processExcluder.Replace(processExcluder)
+
+	pod := fakes.Pod(
+		fakes.WithNamespace("test-ns-excluded"),
+		fakes.WithName("test-name"),
+	)
+	podGVK := schema.GroupVersionKind{Group: "", Version: "v1", Kind: "Pod"}
+
+	unstructuredPod, err := runtime.DefaultUnstructuredConverter.ToUnstructured(pod)
+	require.NoError(t, err)
+	require.NoError(t, cm.AddObject(ctx, &unstructured.Unstructured{Object: unstructuredPod}))
+
+	// test that pod from excluded namespace is not added to the cache
+	cfClient, ok := cm.cfClient.(*FakeCfClient)
+	require.True(t, ok)
+	require.False(t, cfClient.HasGVK(pod.GroupVersionKind()))
+	require.False(t, cfClient.Contains(map[CfDataKey]interface{}{{Gvk: podGVK, Key: "default/config-test-1"}: nil}))
+}
+
+// TestCacheManager_cfClient_errors tests that the cache manager responds to errors from the cfClient client.
+func TestCacheManager_cfClient_errors(t *testing.T) {
+	cm, ctx := unitCacheManagerForTest(t)
+	cfClient, ok := cm.cfClient.(*FakeCfClient)
+	require.True(t, ok)
+	cfClient.SetErroring(true) // This will cause AddObject, RemoveObject to err
+
+	pod := fakes.Pod(
+		fakes.WithNamespace("test-ns"),
+		fakes.WithName("test-name"),
+	)
+	unstructuredPod, err := runtime.DefaultUnstructuredConverter.ToUnstructured(pod)
+	require.NoError(t, err)
+	cm.watchedSet.Add(pod.GroupVersionKind())
+
+	// test that cm bubbles up the errors
+	require.ErrorContains(t, cm.AddObject(ctx, &unstructured.Unstructured{Object: unstructuredPod}), "test error")
+	require.ErrorContains(t, cm.RemoveObject(ctx, &unstructured.Unstructured{Object: unstructuredPod}), "test error")
+}
+
+// TestCacheManager_AddSource tests that we can modify the gvk aggregator and watched set when adding a new source.
+func TestCacheManager_AddSource(t *testing.T) {
```

Comment:
I'll push for table driven testing here as well.  This method of testing basically functions as an integration test with multiple use-cases in a row.  If someone breaks part of this functionality, the granularity of the failure is "AddSource is broken".  It makes it unclear what the failure is, and it's difficult to extend should the details of UpsertSource change in the future.

You could get the same coverage by making a simpler CacheManager object that's instantiated with a GVKAggregator and a watch set.  Then, you could make a table of tests that define before and after states for those components, since they are what's effected by calls to UpsertSource.  You could also define a list of sources that you want to be added via UpsertSource (which allows you to specify duplicates so that you can test the overwrite of a single source).

Later extensions of the code will benefit from this easily extensible paradigm where differences between before and after states are clearly defined.